dev/wlp.lib.extract/src/wlp/lib/extract/SelfExtractor.java

```diff
@@ -518,85 +518,85 @@ public ReturnCode installMissingRequiredFeatures(File outputDir, ReturnCode rc)
      * matches method must return either {@link ProductMatch#NOT_APPLICABLE} or {@link ProductMatch#MATCHED} for every properties file for this method to return
      * {@link ReturnCode#OK}.
      *
-     * @param outputDir Where the product is being installed to
+     * @param outputDir      Where the product is being installed to
      * @param productMatches The list of {@link ProductMatch} objects that need to be satisfied to the current install
      * @return A {@link ReturnCode} indicating if this was successful or not
      */
     public static ReturnCode validateProductMatches(File outputDir, List productMatches) {
-        boolean dirExists;
-        dirExists = outputDir.exists();
-        if (dirExists) {
-            File f = new File(outputDir, "lib/versions");
-            File[] files = f.listFiles();
+
+        List<Properties> props_list;
+        try {
+            props_list = getLibertyProperties(outputDir);
+        } catch (IOException e) {
+            System.err.println(e.getMessage());
+            return new ReturnCode(ReturnCode.BAD_OUTPUT);
+        }
+
+        for (Properties props : props_list) {
+            ReturnCode rc = matchLibertyProperties(productMatches, props);
+            if (rc != ReturnCode.OK) {
+                return rc;
+            }
+        }
+        return ReturnCode.OK;
+
+    }
+
+    /**
+     * @param productMatches
+     * @param props
+     */
+    protected static ReturnCode matchLibertyProperties(List productMatches, Properties props) {
```

Comment:
I dislike returning in methods in multiple areas. I suggest only ever returning once because it's much easier to debug. My suggestion would be to change this to something like:
```
protected static ReturnCode matchLibertyProperties(List productMatches, Properties props) {
        Iterator matches = productMatches.iterator();
        ReturnCode rc = ReturnCode.OK;
        while (matches.hasNext() && rc.getCode() == 0) {
            ProductMatch match = (ProductMatch) matches.next();
            int result = match.matches(props);
            if (result != ProductMatch.NOT_APPLICABLE) {
                String currentLicenseType = props.getProperty("com.ibm.websphere.productLicenseType");
                String currentInstallType = props.getProperty("com.ibm.websphere.productInstallType");
                if (result == ProductMatch.INVALID_VERSION || result == ProductMatch.INVALID_EDITION) {
                    List applicableEditions = InstallUtils.getEditionNameFromList(match.getEditions());
                    String currentEdition = InstallUtils.getEditionName(props.getProperty("com.ibm.websphere.productEdition"));
                    String currentVersion = props.getProperty("com.ibm.websphere.productVersion");
                    if (result == ProductMatch.INVALID_VERSION) {
                        rc = new ReturnCode(ReturnCode.BAD_OUTPUT, "invalidVersion", currentVersion, match.getVersion(), currentEdition, applicableEditions, currentLicenseType);
                    } else {
                        rc = new ReturnCode(ReturnCode.BAD_OUTPUT, "invalidEdition", currentEdition, applicableEditions, currentVersion, match.getVersion(), currentLicenseType);
                    }
                } else if (result == ProductMatch.INVALID_INSTALL_TYPE) {
                    rc = new ReturnCode(ReturnCode.BAD_OUTPUT, "invalidInstallType", currentInstallType, match.getInstallType());
                } else if (result == ProductMatch.INVALID_LICENSE) {
                    rc = new ReturnCode(ReturnCode.BAD_OUTPUT, "invalidLicense", currentLicenseType, match.getLicenseType());
                }
            }
        }
        return rc;
    }
```

Also I would add a static int to ReturnCode like 
```
    public static final int OK_INT = 0;
```

So that you can use a loop invariant that looks more proper:

```
 while (matches.hasNext() && rc.getCode() == ReturnCode.OK_INT
```